server/rest/src/main/java/org/infinispan/rest/server/InfinispanRequest.java

```diff
@@ -0,0 +1,117 @@
+package org.infinispan.rest.server;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Optional;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.QueryStringDecoder;
+import io.netty.handler.codec.http2.HttpConversionUtil;
+
+public class InfinispanRequest {
+
+   private final FullHttpRequest request;
+   private final Optional<String> streamId;
+   private final String cacheName;
+   private final Optional<String> key;
+   private final URI uri;
+
+   public InfinispanRequest(FullHttpRequest request) {
+      this.request = request;
+      this.uri = URI.create(request.uri());
+      streamId = Optional.ofNullable(request.headers().get(HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text()));
+
+      //we are parsing /rest/cacheName/key
+      //not that the first one is empty!
+      String[] pathParams = this.uri.getPath().split("/");
+      switch (pathParams.length) {
+         case 4:
+            cacheName = pathParams[2];
+            key = Optional.of(pathParams[3]);
+            break;
+         case 3:
+            cacheName = pathParams[2];
+            key = Optional.empty();
+            break;
+         default:
+            throw new IllegalArgumentException("Wrong number of Path Parameters");
+      }
+   }
+
+   public Optional<String> getStreamId() {
+      return streamId;
+   }
+
+   public FullHttpRequest getRawRequest() {
+      return request;
+   }
+
+   public String getCacheName() {
+      return cacheName;
+   }
+
+   public Optional<String> getKey() {
+      return key;
+   }
+
+   public Optional<Boolean> getUseAsync() {
+      String performAsync = request.headers().get("performAsync");
+      if ("true".equals(performAsync)) {
+         return Optional.of(Boolean.TRUE);
+      }
+      return Optional.empty();
+   }
+
+   public Optional<String> getMediaType() {
+      return Optional.ofNullable(request.headers().get("Content-Type"));
+   }
+
+   public Optional<byte[]> data() {
+      if(request.content() != null) {
+         ByteBuf content = request.content();
+         if(content.hasArray()) {
+            return Optional.of(content.array());
+         } else {
+            return Optional.of(content.copy().array());
+         }
+      }
+      return Optional.empty();
+   }
+
+   public Optional<Long> getTimeToLiveSeconds() {
+      try {
+         return Optional.of(Long.parseLong(request.headers().get("timeToLiveSeconds")));
+      } catch (NumberFormatException nfe) {
+         return Optional.empty();
```

Comment:
It's there but in a different place: https://github.com/infinispan/infinispan/pull/5094/files#diff-c1c4ddf87770141381278e928f0bc5b0R57